modules/world/world.hpp

```diff
@@ -115,14 +117,15 @@ class World : public commons::BaseType {
 
 typedef std::shared_ptr<world::World> WorldPtr;
 
-inline World *World::Clone() const {
-  World *new_world = new World(*this);
+inline WorldPtr World::Clone() const {
+  WorldPtr new_world = std::make_shared<World>(*this);
   new_world->clear_all();
   for (auto agent = agents_.begin(); agent != agents_.end(); ++agent) {
-    new_world->add_agent(AgentPtr(agent->second->Clone()));
+    new_world->add_agent(
+      std::dynamic_pointer_cast<Agent>(agent->second->Clone()));
   }
   for (auto object = objects_.begin(); object != objects_.end(); ++object) {
-    new_world->add_object(ObjectPtr(object->second->Clone()));
+    new_world->add_object(object->second->Clone());
```

Comment:
I would prefer if we defined object as an abstract class and derive agent and obstacle from it. Maybe we should put that into an issue?

Reply:
An agent is already derived from an object. For me this makes sense.
I do not think that we need to define an object as an abstract class as there are no functions that will be shared, such as "Plan"